### PR TITLE
fix hang issue exposed when intensive traffic in dense network

### PR DESCRIPTION
**Pull request by librasungirl**
The main intention of this PR is to resolve device hang issue.

Mainly include:
- fix hang issue exposed when intensive traffic in dense network.
- mSockAddr should not be reset when send response for CoAP CON message.
- other code refinements.

@jwhui, please help to review it, thanks!

### CODE REVIEW

File: src/core/net/ip6_mpl.hpp [hunk 1]
```diff
@@ -518,7 +518,7 @@ class Mpl
         kNumSeedEntries = OPENTHREAD_CONFIG_MPL_SEED_SET_ENTRIES,
         kSeedEntryLifetime = OPENTHREAD_CONFIG_MPL_SEED_SET_ENTRY_LIFETIME,
         kSeedEntryLifetimeDt = 1000,
-        kDataMessageInterval = 64
+        kDataMessageInterval = 200
```
Comment (LuDuda) [inline src/core/net/ip6_mpl.hpp]: It violates Thread Specification. If with 64ms it works badly, maybe we should point it out in the Thread Group.

File: src/core/thread/mle_constants.hpp [hunk 2]
```diff
@@ -126,7 +126,7 @@ enum
 enum
 {
     kMplChildDataMessageTimerExpirations  = 0, ///< Number of MPL retransmissions for Children.
-    kMplRouterDataMessageTimerExpirations = 2, ///< Number of MPL retransmissions for Routers.
+    kMplRouterDataMessageTimerExpirations = 1, ///< Number of MPL retransmissions for Routers.
```
Comment (LuDuda) [inline src/core/thread/mle_constants.hpp]: It violates Thread Specification. If it works badly, maybe we should point it out in the Thread Group.

File: src/core/net/ip6_mpl.hpp (hunk 1 above)
Comment (librasungirl) [inline src/core/net/ip6_mpl.hpp]: Thanks @LuDuda for pointing this out. I didn't notice the definition in the spec when analysing. Will recover back to ensure compliant with the spec.

However as far as the case REED_5_2_5 in considered, it involves 18 devices with good connectivity, if following the parameters in the spec, many CCA/NoACK error would be observed on the device side and the retransmission would bring more traffic and increase such error. @jwhui, are there any considerations when defining the parameters DATA_MESSAGE_IMIN/DATA_MESSAGE_IMAX and DATA_MESSAGE_TIMER_EXPIRATIONS in the spec?

File: src/core/thread/mle_constants.hpp (hunk 2 above)
Comment (librasungirl) [inline src/core/thread/mle_constants.hpp]: Thanks @LuDuda. Will recover back.

File: examples/platforms/cc2538/radio.c [hunk 3]
```diff
@@ -394,7 +393,14 @@ void cc2538RadioProcess(otInstance *aInstance)
         else
 #endif
         {
-            otPlatRadioReceiveDone(aInstance, &sReceiveFrame, sReceiveError);
+            // signal MAC layer for each received frame if promiscous if enabled
```
Comment (jwhui) [inline examples/platforms/cc2538/radio.c]: nit: `if enabled` -> `is enabled`

**Pull request merged**